llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -521,16 +523,77 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
         //          %1, i32 %2), !SYCL_SPEC_CONST_SYM_ID !23
         // !23 = {!"string-id-2", i32 3, i32 4}
       } else {
-        // 2a. Spec constant must be resolved at compile time - just replace
-        // the intrinsic with default C++ value for the spec constant type.
-        Value *Default = getDefaultCPPValue(SCTy);
-        if (IsComposite) {
-          // __sycl_getCompositeSpecConstant returns through argument, so, the
-          // only thing we need to do here is to store into a memory pointed by
-          // that argument
-          new StoreInst(Default, CI->getArgOperand(0), CI);
+        // 2a. Spec constant must be resolved at compile time - replace the
+        // intrinsic with the actual value for spec constant.
+
+        bool Is2020Intrinsic =
+            F.getName().startswith(SYCL_GET_SCALAR_2020_SPEC_CONST_VAL) ||
+            F.getName().startswith(SYCL_GET_COMPOSITE_2020_SPEC_CONST_VAL);
+
+        if (Is2020Intrinsic) {
+          // Handle SYCL2020 version of intrinsic - replace it with a pointer to
+          // specialization constant value.
+          // A pointer to a single RT-buffer with all the values of
+          // specialization constants is passed as a 3rd argument of intrinsic.
+          Value *RTBuffer =
+              IsComposite ? CI->getArgOperand(3) : CI->getArgOperand(2);
+
+          // Add the string literal to a "spec const string literal ID" ->
+          // "offset" map, uniquing the integer offsets if this is new
+          // literal.
+          auto Ins = OffsetMap.insert(std::make_pair(SymID, Offset));
+          bool IsNewSpecConstant = Ins.second;
+          auto CurrentOffset = Ins.first->second;
+          if (IsNewSpecConstant) {
+            if (IsComposite) {
+              // When handling elements of a structure, we do not use manually
+              // calculated offsets (which are sum of sizes of all previously
+              // encountered elements), but instead rely on data provided for us
+              // by DataLayout, because the structure can be unpacked, i.e.
+              // padded in order to ensure particular alignment of its elements.
+              auto *StructTy = cast<StructType>(
+                  CI->getArgOperand(0)->getType()->getPointerElementType());
+              const StructLayout *SL =
+                  M.getDataLayout().getStructLayout(StructTy);
+              Offset += SL->getSizeInBytes();
+            } else
+              Offset += SCTy->getScalarSizeInBits() / __CHAR_BIT__;
```

Comment:
Why `CHAR_BIT` is used with underscores? [cppreference](https://en.cppreference.com/w/cpp/types/climits) says it is defined just as `CHAR_BIT` in `<climits>` header

Reply:
Good point, will fix.